core/src/main/scala/pureconfig/BasicReaders.scala

```diff
@@ -66,11 +67,19 @@ trait PrimitiveReaders {
   */
 trait JavaEnumReader {
 
-  implicit def javaEnumReader[A <: java.lang.Enum[A]](implicit tag: ClassTag[A]): ConfigReader[A] =
+  private[pureconfig] def baseJavaEnumReader[A <: java.lang.Enum[A]](
+      transformValue: String => String
+  )(implicit tag: ClassTag[A]): ConfigReader[A] =
     ConfigReader.fromString(catchReadError(s => {
       val enumClass = tag.runtimeClass.asInstanceOf[Class[A]]
-      Enum.valueOf(enumClass, s)
+      Enum.valueOf(enumClass, transformValue(s))
     }))
+
+  implicit def javaEnumReader[A <: java.lang.Enum[A]](implicit tag: ClassTag[A]): ConfigReader[A] =
+    baseJavaEnumReader[A](identity)
+
+  private[pureconfig] def mixedCaseJavaEnumReader[A <: java.lang.Enum[A]](implicit tag: ClassTag[A]): ConfigReader[A] =
```

Comment:
So I realized that there is already an `implicit def javaEnumReader` however the `Enum.valueOf` function that is used for Java readers typically only work with upper case values since that is the standard for Java enums.

When dealing with typesafe config howerver, its not typical to use uppercase values, i.e. in the context of this PR you would write something like

```
my-chrono-unit=millis
```

and not
```
my-chrono-unit=MILLIS
```

Hence why I created a private `mixedCaseJavaEnumReader` which also handles cases such as `Millis` and `millis`. There is an argument that the default `javaEnumReader` should have been a mixed case one rather than uppercase however this would break existing code out there so I left it as is. There is also merit in having `mixedCaseJavaEnumReader` as public however I am not sure if you want to expand the trait's namespace like this.

Reply:
I like the idea of adding a public reader for Java enums with a configurable field mapping, but I have a few comments:

- `mixedCaseJavaEnumReader` is probably too specific for your case. For example, using kebab case (`MY_ENUM` -> `my–enum`) is probably more standard than simply having everything lower case. I'd rather have `baseJavaEnumReader` public and use it directly in `chronoUnitReader`;
- We have a special implementation for `String => String` functions with the purpose of `transformValue`, please see `ConfigFieldMapping`. Just like in product/coproduct hints, we should probably use `ConfigFieldMapping` directly instead of a general function (in your specific case, it would be `ConfigFieldMapping(ScreamingSnakeCase, KebabCase)`);
- `BasicReaders` should include only ready-to-use reader instances, not instances that need to be "configured". We have a package `pureconfig.configurable` to hold readers that need to be constructed explicitly. Can you please move your `baseJavaEnumReader` there, renaming it to something like `javaEnumReaderWithMapping`?